daemon/resources/man/razer.conf.5
.\" Generated by scdoc 1.11.3
.\" Complete documentation for this program is not available as a GNU info page
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.nh
.ad l
.\" Begin generated content:
.TH "razer.conf" "5" "2024-07-02"
.PP
.SH NAME
.PP
razer.\&conf - configuration file for the OpenRazer Daemon
.PP
.SH DESCRIPTION
.PP
The \fI~/.\&config/openrazer/razer.\&conf\fR (\fB$XDG_CONFIG_HOME\fR is respected, if set) file configures some startup and runtime options of the \fBopenrazer-daemon\fR(8).\& \fBopenrazer-daemon\fR(8) is able to use configuration files specified in non-standard locations by providing the \fI--config\fR argument.\&
.PP
The configuration file is formatted like a .\&INI file and is parsed by Python'\&s ConfigParser.\&
.PP
.SH EXAMPLE
.PP
An example file should be installed under \fI/usr/share/openrazer/razer.\&conf.\&example\fR.\&
.PP
.SH SECTIONS
.PP
.SS GENERAL
.PP
The first section in the configuration file is a \fB[General]\fR section.\& This section contains miscellaneous options.\&
.PP
\fBverbose_logging\fR \fIbool\fR
.RS 4
This flag specifies if the daemon is to output detailed logging information.\& This value acts the same as if \fB-v\fR or \fB--verbose\fR was passed at the command line.\&
.PP
.RE
.SS STARTUP
.PP
The \fB[Startup]\fR section in the configuration file contains values to be used during startup, for example it can decide if syncing effects will be active when started.\&
.PP
\fBsync_effects_enabled\fR \fIbool\fR
.RS 4
This flag specifies if the effects syncing logic is active when the daemon is started, not having to wait for the user to activate them.\&
.PP
.RE
\fBdevices_off_on_screensaver\fR \fIbool\fR
.RS 4
This flag specifies if the functionality to turn off razer devices when the screensaver is activated is active when the daemon starts.\&
.PP
.RE
\fBbattery_notifier\fR \fIbool\fR
.RS 4
This flag specifies that notifications should be shown regularly with the remaining battery level.\&
.PP
.RE
\fBbattery_notifier_freq\fR \fIint\fR
.RS 4
This flag specifies how often the notification with the remaining battery level should be shown.\&
.PP
.RE
\fBbattery_notifier_percent\fR \fIint\fR
.RS 4
This flag specifies what battery percentage the device should reach before sending notifications.\&
.PP
.RE
\fBmouse_dpi_notifier\fR \fIbool\fR
.RS 4
This flag specifies that notifications should be shown regularly when mouse DPI changes.\&
.P
.RE
\fBmouse_dpi_notifier_freq\fR \fIint\fR
.RS 4
This flag specifies how often, in milliseconds, the notification for mouse DPI changes should be shown.\&
.P
.RE
\fBrestore_persistence\fR \fIbool\fR
.RS 4
This flag specifies whether effects saved in persistence.\&conf should be applied when the daemon starts.\&
.PP
.RE
\fBpersistence_dual_boot_quirk\fR \fIbool\fR
.RS 4
This flag specifies whether the persisted effect is applied twice to address a quirk with newer devices that were last changed by Razer Synapse on Windows.\& Only applicable with restore_persistence = True.\&
.PP
.RE
.SH SEE ALSO
.PP
\fBopenrazer-daemon\fR(8)


debian/changelog
openrazer (3.9.0) noble; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.9.0

 -- Luca Weiss <luca@lucaweiss.eu>  Sun, 06 Oct 2024 11:40:51 +0200

openrazer (3.8.0) jammy; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.8.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 14 Apr 2024 11:34:24 +0200

openrazer (3.7.0) jammy; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.7.0

 -- Luca Weiss <luca@z3ntu.xyz>  Wed, 01 Nov 2023 20:47:56 +0200

openrazer (3.6.1) jammy; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.6.1

 -- Luca Weiss <luca@z3ntu.xyz>  Sat, 13 May 2023 17:02:29 +0200

openrazer (3.6.0) jammy; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.6.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 30 Apr 2023 14:23:43 +0200

openrazer (3.5.1) jammy; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.5.1

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 04 Dec 2022 12:48:59 +0100

openrazer (3.5.0) jammy; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.5.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 20 Nov 2022 15:01:08 +0100

openrazer (3.4.0) jammy; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.4.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 31 Jul 2022 18:00:44 +0200

openrazer (3.3.0) focal; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.3.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sat, 09 Apr 2022 15:57:47 +0200

openrazer (3.2.0) focal; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.2.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 12 Dec 2021 15:58:06 +0100

openrazer (3.1.0) focal; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.1.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 15 Aug 2021 11:42:45 +0200

openrazer (3.0.1) bionic; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.0.1

 -- Luca Weiss <luca@z3ntu.xyz>  Sat, 10 Apr 2021 11:20:13 +0200

openrazer (3.0.0) bionic; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v3.0.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 21 Mar 2021 17:26:04 +0100

openrazer (2.9.0) bionic; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.9.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 08 Nov 2020 17:01:11 +0100

openrazer (2.8.0) bionic; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.8.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sat, 02 May 2020 13:20:33 +0200

openrazer (2.7.0) bionic; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.7.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sat, 11 Jan 2020 11:18:19 +0100

openrazer (2.6.0) bionic; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.6.0

 -- Luca Weiss <luca@z3ntu.xyz>  Mon, 29 Jul 2019 16:22:52 +0200

openrazer (2.5.0) bionic; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.5.0

 -- Luca Weiss <luca@z3ntu.xyz>  Thu, 28 Mar 2019 19:06:54 +0100

openrazer (2.4.0) xenial; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.4.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 21 Oct 2018 18:21:27 +0200

openrazer (2.3.1) xenial; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.3.1

 -- Luca Weiss <luca@z3ntu.xyz>  Wed, 11 Jul 2018 17:09:23 +0200

openrazer (2.3.0) xenial; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.3.0

 -- Luca Weiss <luca@z3ntu.xyz>  Sun, 01 Apr 2018 15:49:19 +0200

openrazer (2.2.2) xenial; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.2.2

 -- Luca Weiss <luca@z3ntu.xyz>  Tue, 27 Feb 2018 20:45:43 +0100

openrazer (2.2.1) xenial; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.2.1

 -- Luca Weiss <luca@z3ntu.xyz>  Fri, 23 Feb 2018 21:51:47 +0100

openrazer (2.2.0) xenial; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.2.0

 -- Luca Weiss <luca@z3ntu.xyz>  Tue, 26 Dec 2017 17:29:26 +0100

openrazer (2.1.1) xenial; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.1.1

 -- Luca Weiss <luca@z3ntu.xyz>  Sat, 18 Nov 2017 12:35:27 +0100

openrazer (2.1-0ubuntu1) xenial; urgency=high

  * Changelog available at https://github.com/openrazer/openrazer/releases/tag/v2.1

 -- Luca Weiss <luca@z3ntu.xyz>  Tue, 14 Nov 2017 21:15:36 +0100

openrazer (2.0.0-0ubuntu1) xenial; urgency=high

  * Renamed to OpenRazer.

 -- Luke Horwell <luke@ubuntu-mate.org>  Sat, 2 Sep 2017 12:42:00 +0000

razer (1.1.16-0ubuntu1) xenial; urgency=high
  * Added Device - Razer Blade Stealth Mid 2017
  * Added Device - Razer Naga Chroma
  * Added Device - Razer AbyssusV2
  * Added fake driver config - Razer BlackWidow Chroma TE
  * Added fake driver config - Razer BlackWidow Chroma V2
  * Added fake driver config - Razer BlackWidow Original + Alt
  * Added fake driver config - Razer BlackWidow Ultimate 2012, 2013, 2016
  * Added fake driver config - Razer BlackWidow X Chroma, TE
  * Added fake driver config - Razer BlackWidow X Ultimate
  * Added fake driver config - Razer Blade Late 2016
  * Added fake driver config - Razer Blade Pro Late 2016
  * Added fake driver config - Razer Blade QHD
  * Added fake driver config - Razer Blade Stealth Late 2016
  * Added fake driver config - Razer Blade Stealth Mid 2017

scripts/wireshark/razer.lua
-- Thanks to karlp for making a nice lua protocol example for USB https://github.com/karlp/swopy
--
-- Copyright 2015 Terri Cain <terri@dolphincorp.co.uk>
-- To install put in ~/.wireshark/plugins/FILENAME.lua
-- or ran with wireshark -X lua_script:FILENAME.lua
--
-- Could of binded it to the Razer USB Vendor:Product ID but when wiresharking USB comms to virtualbox
-- wireshark can miss the VID,PID declaration and then the dissector wouldn't apply so just use it when
-- dissecting razer traffic, don't use it all the time as it'll apply to all usb-control requests

razer_proto = Proto("razer", "Razer Protocol")

local req_types = {
    [0x00] = "REQUEST",
    [0x02] = "RESPONSE",
}

local f = razer_proto.fields
f.f_status = ProtoField.uint8("razer.status", "Request Type", base.HEX, req_types)                  -- 1b
f.f_id = ProtoField.uint8("razer.id", "ID", base.HEX)                                               -- 1b Possibly I2C range
f.f_remaining_packets = ProtoField.uint16("razer.remaining_packets", "Remaining Packets", base.DEC) -- 2b
f.f_protocol_type = ProtoField.uint8("razer.protocol_type", "Protocol Type", base.HEX)              -- 1b
f.f_number_parameter_bytes = ProtoField.uint8("razer.num_params", "Number of parameters", base.HEX) -- 1b Number of bytes after command byte
f.f_command_class = ProtoField.uint8("razer.command_class", "Command Class", base.HEX)              -- 1b Command byte
f.f_command_id = ProtoField.uint8("razer.command_id", "Command ID", base.HEX)                       -- 1b Sub Command byte
f.f_command_params = ProtoField.bytes("razer.command_params", "Command parameters", base.SPACE)     -- 80b Params
f.f_crc = ProtoField.uint8("razer.crc", "CRC", base.HEX)                                            -- 1b CRC checksum
f.f_end_marker = ProtoField.uint8("razer.end", "End", base.HEX)                                     -- 1b End marker


local f_usb_ep_num = Field.new("usb.endpoint_address.number")    -- should be 0
local f_usb_ep_dir = Field.new("usb.endpoint_address.direction") -- 1 IN, 0 OUT
local f_data_len = Field.new("usb.data_len") -- should be 90
local f_urb_len = Field.new("usb.urb_len") -- should be 90

local function getstring(fi)
    local ok, val = pcall(tostring, fi)
    if not ok then val = "(unknown)" end
    return val
end

-- write32 doesn't have a response on the in endpoint, it tweaks decoding on the _out_ endpoint
local responses = {
    NOTSET = 1, READMEM32 = 2, GENERIC = 3, READDEBUG = 4,
    WRITEMEM32 = 5,
    TRACECOUNT = 6
}

local expected = responses.NOTSET

function razer_proto.dissector(buffer, pinfo, tree)

    --[[This was very helpful for working out the field names I could use with Field.new()
    local fields = { all_field_infos() }
    for ix, finfo in ipairs(fields) do
        print(string.format("ix=%d, finfo.name = %s, finfo.value=%s", ix, finfo.name, getstring(finfo)))
    end
    print("\n")--]]

    local data_length = f_data_len()
    local data_direction = f_usb_ep_dir()
    local urb_length = f_urb_len()


    if (data_length.value == 90 and urb_length.value == 90) then
        pinfo.cols["protocol"] = "Razer"

        -- seek to the last 90 bytes
        local offset = buffer:len() - 90
        local t_razer = tree:add(razer_proto, buffer())

        if (data_direction.value == 0) then
            pinfo.cols["info"]:append("Request")
        else
            pinfo.cols["info"]:append("Response")
        end

        -- Add REQ,RES byte
        t_razer:add(f.f_status, buffer(offset, 1))
        offset = offset + 1
        -- Add ID
        t_razer:add(f.f_id, buffer(offset, 1))
        offset = offset + 1
        -- Add Remaining packets
        t_razer:add(f.f_remaining_packets, buffer(offset, 2))
        offset = offset + 2
        -- Add Protocol Type
        t_razer:add(f.f_protocol_type, buffer(offset, 1))
        offset = offset + 1
        -- Add Number of parameters bytes
        local num_params = buffer(offset, 1):le_uint()
        t_razer:add(f.f_number_parameter_bytes, buffer(offset, 1)) -- specified buffer here instead of num params so it highlights in wireshark
        offset = offset + 1
        -- Add Command class
        t_razer:add(f.f_command_class, buffer(offset, 1))
        offset = offset + 1
        -- Add Command ID
        t_razer:add(f.f_command_id, buffer(offset, 1))
        offset = offset + 1
        -- Add command params
        t_razer:add(f.f_command_params, buffer(offset, num_params))
        offset = offset + 80 -- skip to the end
        -- Add CRC
        t_razer:add(f.f_crc, buffer(offset, 1))
        offset = offset + 1
        -- Add end
        t_razer:add(f.f_end_marker, buffer(offset, 1))
        offset = offset + 1
    end
end

usb_table = DissectorTable.get("usb.control")
usb_table:add(0xffff, razer_proto)
